tensorflow/python/ops/embedding_ops.py

```diff
@@ -887,34 +928,50 @@ def safe_embedding_lookup_sparse(embedding_weights,
-        if original_rank_dim is None else original_rank_dim)
-    sparse_ids = sparse_ops.sparse_reshape(sparse_ids, [
-        math_ops.reduce_prod(
-            array_ops.slice(original_shape, [0], [original_rank - 1])),
-        array_ops.gather(original_shape, original_rank - 1)
-    ])
-    if sparse_weights is not None:
-      sparse_weights = sparse_tensor.SparseTensor(sparse_ids.indices,
-                                                  sparse_weights.values,
-                                                  sparse_ids.dense_shape)
-
-    # Prune invalid ids and weights.
-    sparse_ids, sparse_weights = _prune_invalid_ids(sparse_ids, sparse_weights)
-    if combiner != "sum":
-      sparse_ids, sparse_weights = _prune_invalid_weights(
-          sparse_ids, sparse_weights)
-
-    # Fill in dummy values for empty features, if necessary.
-    sparse_ids, is_row_empty = sparse_ops.sparse_fill_empty_rows(
+    if isinstance(sparse_ids, sparse_tensor.SparseTensor):
+      # Reshape higher-rank sparse ids and weights to linear segment ids.
+      original_shape = sparse_ids.dense_shape
+      original_rank_dim = tensor_shape.dimension_value(
+          sparse_ids.dense_shape.get_shape()[0])
+      original_rank = (
+          array_ops.size(original_shape)
+          if original_rank_dim is None else original_rank_dim)
+      sparse_ids = sparse_ops.sparse_reshape(sparse_ids, [
+          math_ops.reduce_prod(
+              array_ops.slice(original_shape, [0], [original_rank - 1])),
+          array_ops.gather(original_shape, original_rank - 1)
+      ])
+      if sparse_weights is not None:
+        sparse_weights = sparse_tensor.SparseTensor(sparse_ids.indices,
+                                                    sparse_weights.values,
+                                                    sparse_ids.dense_shape)
+
+      # Prune invalid ids and weights.
+      sparse_ids, sparse_weights = \
+                            _prune_invalid_ids(sparse_ids, sparse_weights)
+      if combiner != "sum":
+        sparse_ids, sparse_weights = _prune_invalid_weights(
+            sparse_ids, sparse_weights)
+
+      # Fill in dummy values for empty features, if necessary.
+      sparse_ids, is_row_empty = sparse_ops.sparse_fill_empty_rows(
         sparse_ids, default_id or 0)
-    if sparse_weights is not None:
-      sparse_weights, _ = sparse_ops.sparse_fill_empty_rows(sparse_weights, 1.0)
+      if sparse_weights is not None:
+        sparse_weights, _ = sparse_ops.sparse_fill_empty_rows(
+            sparse_weights, 1.0)
+
+    else:
+      # Prune invalid ids and weights.
+      sparse_ids, sparse_weights = _prune_invalid_ids_ragged(sparse_ids,
+                                                             sparse_weights)
+      if combiner != "sum":
+        sparse_ids, sparse_weights = _prune_invalid_weights_ragged(
+            sparse_ids, sparse_weights)
+      sparse_ids, is_row_empty = ragged_array_ops.fill_empty_rows(
```

Comment:
Do we know what errors could be raised in `embedding_lookup_sparse` if we don't fill empty rows?

Reply:
`embedding_lookup_sparse` currently returns zeros in rows that are empty in the sparse input.